lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+Converter::toCABI(const model::RawFunctionType &Function,
+                  TupleTree<model::Binary> &Binary) const {
+  revng_log(Log, "Converting a `RawFunctionType` to `CABIFunctionType`.");
+  revng_log(Log, "Function:\n" << serializeToString(Function));
+
+  if (!ABI.checkCompatibility(Function.Arguments(), Function.ReturnValues())) {
+    revng_log(Log,
+              "FAIL: the function is not compatible with `"
+                << model::ABI::getName(ABI.ABI()) << "`.");
+    return std::nullopt;
+  }
+
+  // A single place to conveniently store all the types that need to be added
+  // to the model in order for the new function prototype to be valid.
+  auto M = manageTypeDependencies(*Binary);
+
+  // Convert register arguments first.
+  auto RegArguments = M.tryConvertingRegisterArguments(Function.Arguments());
+  if (RegArguments == std::nullopt) {
+    revng_log(Log, "FAIL: Register argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // Then stack arguments.
+  auto Stack = M.tryConvertingStackArguments(Function.StackArgumentsType(),
+                                             RegArguments->size());
+  if (Stack == std::nullopt) {
+    revng_log(Log, "FAIL: Stack argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // And the return type.
+  auto ReturnType = M.tryConvertingReturnValue(Function.ReturnValues());
+  if (ReturnType == std::nullopt) {
+    revng_log(Log, "FAIL: Return value conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // If all the partial conversions passed, the conversion is guaranteed to be
+  // successful. All the `TypeDependencies` can now be recorded into the
+  // model.
+  M.commit();
+
+  // With all the dependencies out of the way, new `CABIFunctionType` can now
+  // be created,
+  auto [NewType, NewTypePath] = Binary->makeType<model::CABIFunctionType>();
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
+
+  // To finish up the conversion, remove all the references to the old type by
+  // carefully replacing them with references to the new one.
+  return replaceAllUsesWith(Function.key(), NewTypePath, Binary);
+}
+
+using TDManager = TypeDependencyManager;
+template<typename RT>
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingRegisterArguments(const SortedVector<RT> &Used) {
+  llvm::SmallVector<model::Register::Values, 8> ArgumentRegisters;
+
+  // Rely onto the register state deduction to make sure no "holes" are
+  // present in-between the argument registers.
```

Comment:
Sorry, I wasn't clear. I'm suggesting to fail the translation in case there's a single all (if we're in strict conversion mode, i.e., `not ProduceCompatible`).

Maybe `bool Strict` would be a better flag name.

Reply:
I feel like going that way we might give up on converting a whole load of functions that just have an unused argument. And it feels really sad.

I get that there might be cases where we have no choice. But if there's even a chance that limiting the conversion so hard can be avoided, I think that would be a better alternative.

Reply:
Consider that:
* EFA should already massage the prototype to fill in the holes IIRC, so this is redundant in the "standard" pipeline;
* I'm suggesting to have a flag to be stricter, but I'm fine with having it turned off by default; it's just for testing and for having two operation modes, one that produces a `CABIFunctionType` *compatible* with the original `RawFunctionType` (non-strict mode) and one that either produces a `CABIFunctionType` that would be *identical* when going back to `RawFunctionType` or it fails (strict-mode).

Reply:
1. But we also don't have the way to make sure this analysis depends on that one (not that we'd want to). I'm pretty sure this is often going to get run on non-EFA-ed models (since it's a pure analysis, and I feel like the users are going to be eager to run it - no-one wants to see `Raw` functions).

2. I'm not against it per se. I'm just concerned that it's extra logic to maintain (public-interface-affecting at that) for a pretty minor benefit. (For some reason I'm feeling like your influence is rubbing off on me, not that it's a bad thing `:D`)

Either way, we can try and see how it looks. No harm to be done.

Reply:
I understand there's extra cost here, but the benefit is that we're forcing ourselves to neatly distinguish between "destructive" and "non-destructive" parts of the conversion.

If you don't feel like doing this, please specify in the docs that we guarantee that if you do `Raw -> CABI -> Raw` the original and final `Raw` will never be identical but just "compatible". That's fine too.

Reply:
> we have to throw away like 70% of all the possible conversions

It's an option, not the default. We might decide to enable this for a pipeline going through EFA (which already does fill the holes) to test things.
Failing conversion is perfectly fine. A failure it's one more opportunity for us to see if we're doing something wrong upstream from that.

> because we loose type information

Yeah, identical for what concerns everything except types (but including their size?).

At this point, if we want to fill the holes here. Why fill them in EFA?

In `DetectABI::applyABIDeductions` we have:

```c++
    if (EnforceABIConformance) {
      ResultMap = enforceRegisterStateDeductions(StateMap,
                                                 Binary->DefaultABI());
    } else {
      ResultMap = tryApplyRegisterStateDeductions(StateMap,
                                                  Binary->DefaultABI());
    }
```

If we fill in the holes in `toCABI`, we could drop all the logic for `EnforceABIConformance == false` and basically limit `RegisterStateDeductions` to suppress arguments/return values that are not compatible with the given ABI. No filling of holes.

Can you consider this option? I'm not familiar enough with this, but there's clearly a duplication of roles between filling the holes in `toCABI` and filling the holes in `RegisterStateDeductions`. If we don't want to make it optional in `toCABI` we can consider not doing it all in `RegisterStateDeductions`.

Reply:
> We might decide to enable this for a pipeline going through EFA

I'm not sure that's beneficial. Consider that the code EFA uses and the code used here is 100% shared, so by keeping it enabled in that specific case, it just does nothing. Except for maybe verifying that EFA really ran the strict deduction, not the soft one.

> but including their size?

Yep, size is preserved.

> Why fill them in EFA?

Actually, now that you say that, why do we do that in EFA? I always assumed there's something in the pipeline that would benefit from it. Something for each CABI conversion is too late in the pipeline to impact.

What are the advantages of adding extra registers to RFTs? I feel like this is a really good indicator of the fact that the argument in question is never used. We might even want to try and encode that information into the CABI at some point, could be cool to show that explicitly in the decompiled code.

> If we fill in the holes in `toCABI`, we could drop all the logic for `EnforceABIConformance == false` and basically limit `RegisterStateDeductions` to suppress arguments/return values that are not compatible with the given ABI. No filling of holes.

Not exactly, it would still fill stuff. But there's already a really similar `Definition::verifyCompatibility(const RawFunctionType &)` (or something like that, I don't remember the exact name) function that can be easily extended to handle this case as well.

> Can you consider this option?

IMO, if there's no other points in the pipeline that would benefit from holes being filled, I'd rather move it closer to the actual user - to avoid any possibilities of interference in-between. But worse case, running it twice is not much of a problem, considering it's the same code that does nothing if the function already complies.

Reply:
> Consider that the code EFA uses and the code used here is 100% shared

Nice!

> What are the advantages of adding extra registers to RFTs?

Mostly FUD over the possibility of bugs/limitations in EFA. But if an argument is not marked as an argument by EFA (modulo FUD/bugs) it's because it's either not-an-argument or unused. Marking an unused argument as used in `RawFunctionType` does not bring any benefit (except for making life easier for `toCABI`). If `toCABI` does this, I think I'd be happier not to complicate prototypes produced by EFA needlessly.

> We might even want to try and encode that information into the CABI at some point, could be cool to show that explicitly in the decompiled code

Interesting but that would be a pipe, not an analysis. It's something we can infer from the IR. Out of scope.

> IMO, if there's no other points in the pipeline that would benefit from holes being filled, I'd rather move it closer to the actual user - to avoid any possibilities of interference in-between

I think there are no benefits, but let's keep this out of scope.
Please put a comment in `DetectABI` stating that we could drop "filling the holes" since it will be performed in `toCABI` and that in fact we could consider entirely dropping register state deductions and live with extra arguments/return values that will be suppressed later on by `toCABI` (even though this requires a bit more thinking).
Implementing this is out of scope, just put a comment please.

Long discussion but we now have an horizon to consider dropping register state deductions entirely, basically making EFA ABI-independent again, which is not bad at all.

Reply:
:revng_blob:

Reply:
The comment: https://github.com/revng/revng/pull/232/commits/635f9607e0f881e89d7abe39eb23425a5ff63d53.